proof/invariant-abstract/AInvs.thy

```diff
@@ -2060,4 +2061,496 @@ lemma handle_event_preemption_path_schact_is_rct_imp_ct_activatable:
+      apply (rule hoare_weaken_pre)
+       apply (rule schedule_switch_thread_branch_ct_not_idle_imp_ct_released)
+      apply fastforce
+     apply (wpsimp wp: schedule_choose_new_thread_ct_not_idle_imp_ct_released)
+     apply (clarsimp simp: valid_sched_def schedulable_def2 ct_ready_if_schedulable_def
+                           vs_all_heap_simps obj_at_kh_kheap_simps)
+
+    apply (rule_tac Q="\<lambda>_ s. valid_state s \<and>  cur_tcb s" in hoare_post_imp)
+     apply (clarsimp simp: valid_state_def)
+    apply (case_tac action; clarsimp)
+      apply wpsimp
+      apply fastforce
+     apply (rename_tac candidate)
+     apply (rule_tac B="\<lambda>_ s. invs s \<and> scheduler_action s = switch_thread candidate"
+                  in hoare_seq_ext[rotated])
+      apply wpsimp
+     apply (rule hoare_seq_ext_skip, solves wpsimp)+
+     apply (rule hoare_if)
+      apply (wpsimp wp: schedule_choose_new_thread_valid_state_cur_tcb
+                        switch_to_thread_invs thread_get_wp
+                    simp: set_scheduler_action_def schedule_switch_thread_fastfail_def)
+     apply (wpsimp wp: schedule_choose_new_thread_valid_state_cur_tcb
+                       switch_to_thread_invs thread_get_wp
+                 simp: set_scheduler_action_def schedule_switch_thread_fastfail_def)
+    apply (wpsimp wp: schedule_choose_new_thread_valid_state_cur_tcb)
+
+   apply (wpsimp simp: schedule_switch_thread_fastfail_def wp: hoare_drop_imps)
+  apply (wpsimp simp: schedule_switch_thread_fastfail_def wp: hoare_drop_imps)
+  apply fastforce
+  done
+
+crunches activate_thread
+  for cur_sc_offset_ready[wp]: "\<lambda>s :: det_state. cur_sc_offset_ready (consumed_time s) s"
+  and cur_sc_offset_sufficient[wp]: "\<lambda>s :: det_state. cur_sc_offset_sufficient (consumed_time s) s"
+
+lemma ct_not_blocked_imp_ct_not_blocked_on_receive:
+  "ct_not_blocked s \<Longrightarrow> ct_not_blocked_on_receive s"
+  apply (auto simp: ct_in_state_def pred_tcb_at_def obj_at_def)
+  apply (case_tac "tcb_state tcb"; clarsimp)
+  done
+
+lemma ct_not_blocked_imp_ct_not_blocked_on_ntfn:
+  "ct_not_blocked s \<Longrightarrow> ct_not_blocked_on_ntfn s"
+  apply (auto simp: ct_in_state_def pred_tcb_at_def obj_at_def)
+  apply (case_tac "tcb_state tcb"; clarsimp)
+  done
+
+lemma call_kernel_cur_sc_offset_ready_and_sufficient:
+  "\<lbrace>\<lambda>s. cur_sc_offset_ready (consumed_time s) s \<and> cur_sc_offset_sufficient (consumed_time s) s
+        \<and> cur_sc_active s \<and> ct_not_in_release_q s
+        \<and> (ct_running s \<or> ct_idle s) \<and> (e \<noteq> Interrupt \<longrightarrow> ct_running s)
+        \<and> invs s \<and> schact_is_rct s \<and> valid_sched s
+        \<and> current_time_bounded 5 s \<and> valid_machine_time s \<and> consumed_time_bounded s\<rbrace>
+   call_kernel e
+   \<lbrace>\<lambda>_ s :: det_state. cur_sc_offset_ready (consumed_time s) s
+                       \<and> cur_sc_offset_sufficient (consumed_time s) s\<rbrace>"
```

Comment:
It's definitely not worth changing any of this, but out of curiosity do you remember whether it was necessary to prove these together?

Reply:
Definitely not necessary, but it did cut down on a lot of repetition. Sometimes readiness and sufficiency are dealt with separately: for example, in `preemption_point`, an sc that is ready will remain ready as time advances, but might become insufficient as the consumed time increases. But a lot of the time they're dealt with together, and `cur_sc_more_than_ready`, which is a major ingredient of these proofs, has them together